subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java

```diff
@@ -0,0 +1,503 @@
+                    return implementation.getClassLoaderHash() != null
+                        ? implementation.getClassLoaderHash().toByteArray()
+                        : null;
+                }
+            })
+            .orElse(null);
+    }
+
+    @Override
+    public List<byte[]> getActionClassLoaderHashesBytes() {
+        return cachingState.getInputs()
+            .map(new java.util.function.Function<CachingInputs, List<byte[]>>() {
+                @Nullable
+                @Override
+                public List<byte[]> apply(CachingInputs inputs) {
+                    List<ImplementationSnapshot> actionImplementations = inputs.getAdditionalImplementations();
+                    if (actionImplementations.isEmpty()) {
+                        return null;
+                    }
+                    return Lists.transform(actionImplementations, new Function<ImplementationSnapshot, byte[]>() {
+                        @Override
+                        public byte[] apply(ImplementationSnapshot input) {
+                            return input.getClassLoaderHash() == null ? null : input.getClassLoaderHash().toByteArray();
+                        }
+                    });
+                }
+            })
+            .orElse(null);
+    }
+
+    @Nullable
+    @Override
+    public List<String> getActionClassNames() {
+        return cachingState.getInputs()
+            .map(new java.util.function.Function<CachingInputs, List<String>>() {
+                @Nullable
+                @Override
+                public List<String> apply(CachingInputs inputs) {
+                    List<ImplementationSnapshot> actionImplementations = inputs.getAdditionalImplementations();
+                    if (actionImplementations.isEmpty()) {
+                        return null;
+                    }
+                    return Lists.transform(actionImplementations, new Function<ImplementationSnapshot, String>() {
+                        @Override
+                        public String apply(ImplementationSnapshot input) {
+                            return input.getTypeName();
+                        }
+                    });
+                }
+            })
+            .orElse(null);
+    }
+
+    @Nullable
+    @Override
+    public List<String> getOutputPropertyNames() {
+        return cachingState.getInputs()
+            .map(new java.util.function.Function<CachingInputs, List<String>>() {
+                @Nullable
+                @Override
+                public List<String> apply(CachingInputs inputs) {
+                    ImmutableSortedSet<String> outputPropertyNames = inputs.getOutputProperties();
+                    if (outputPropertyNames.isEmpty()) {
+                        return null;
+                    }
+                    // Copy should be a NOOP as this is an immutable sorted set upstream.
+                    return ImmutableSortedSet.copyOf(outputPropertyNames).asList();
```

Comment:
No idea. I guess we can return a `Set` without problems, and remove all of this mess.

Reply:
It is important to be sorted. I don't know what returning a `Set` would solve.

Reply:
No we can't, because... reasons. I removed the `copyOf()` though.

Reply:
> It is important to be sorted. I don't know what returning a `Set` would solve.

It would have been a `SortedSet`, which should have worked, but we seem to have some code explicitly looking for a `List`. :/

Reply:
Well, the interface is directly consumed from the build scan plugin, right?